### PR TITLE
Implement RFC 1210: impl specialization

### PR DESCRIPTION
**Pull request by aturon**
This PR implements [impl specialization](https://github.com/rust-lang/rfcs/pull/1210), 
carefully following the proposal laid out in the RFC.

The implementation covers the bulk of the RFC. The remaining gaps I know of are: 
- no checking for lifetime-dependent specialization (a soundness hole);
- no `default impl` yet;
- no support for `default` with associated consts;

I plan to cover these gaps in follow-up PRs, as per @nikomatsakis's preference.

The basic strategy is to build up a _specialization graph_ during
coherence checking. Insertion into the graph locates the right place
to put an impl in the specialization hierarchy; if there is no right
place (due to partial overlap but no containment), you get an overlap
error. Specialization is consulted when selecting an impl (of course),
and the graph is consulted when propagating defaults down the
specialization hierarchy.

You might expect that the specialization graph would be used during
selection -- i.e., when actually performing specialization. This is
not done for two reasons:
- It's merely an optimization: given a set of candidates that apply,
  we can determine the most specialized one by comparing them directly
  for specialization, rather than consulting the graph. Given that we
  also cache the results of selection, the benefit of this
  optimization is questionable.
- To build the specialization graph in the first place, we need to use
  selection (because we need to determine whether one impl specializes
  another). Dealing with this reentrancy would require some additional
  mode switch for selection. Given that there seems to be no strong
  reason to use the graph anyway, we stick with a simpler approach in
  selection, and use the graph only for propagating default
  implementations.

Trait impl selection can succeed even when multiple impls can apply,
as long as they are part of the same specialization family. In that
case, it returns a _single_ impl on success -- this is the most
specialized impl _known_ to apply. However, if there are any inference
variables in play, the returned impl may not be the actual impl we
will use at trans time. Thus, we take special care to avoid projecting
associated types unless either (1) the associated type does not use
`default` and thus cannot be overridden or (2) all input types are
known concretely.

r? @nikomatsakis

### CODE REVIEW

**Comment by cristicbz** [comment]
:confetti_ball: :balloon: :confetti_ball: 

Regardless which way the RFC goes, this is tremendous work @aturon, big props!

**Comment by aturon** [comment]
Obviously, needs a rebase :)

Also, I forgot to mention, but this passes the full existing test suite as well as a pile of new tests, so it should be in a pretty good state.

**Comment by aturon** [comment]
Regarding the gaps: 

> no `partial impl` yet;

This should be pretty easy to add in a follow-up, given the infrastructure laid by this PR.

> no specialization on inherent impls;

I'm really not sure about this one; right now, inherent impls work pretty independently from our trait infrastructure, so this is potentially a major refactor. It's also one of the more subtle bits of the design. I think we should probably sit on this for a while.

> no support for `default` with associated consts;

Should be easy to add; can probably sneak it into this PR.

> no support for specializing based on trait hierarchies (super traits).

I'm not actually sure why this isn't working yet; somehow the selection infrastructure is failing to elaborate supertrait relationships. Probably an easy fix?

**Comment by dylanede** [comment]
I've just tried out this PR, and there seems to be a problem with associated type expansion. With this code:

```rust
#![feature(specialization)]

use std::marker::PhantomData;

struct Undefined;

struct Nil;
struct Cons<Head, Tail>(PhantomData<(Head, Tail)>);

trait Lookup_<Key> {
    type Value;
}

type Lookup<Key, L> = <L as Lookup_<Key>>::Value;

impl<Key, T> Lookup_<Key> for T {
    default type Value = Undefined;
}

impl<Key, Head, Tail> Lookup_<Key> for Cons<Head, Tail> {
    default type Value = <Tail as Lookup_<Key>>::Value;
}

impl<Key, Value, Tail> Lookup_<Key> for Cons<(Key, Value), Tail> {
    type Value = Value;
}

trait Eval_<Env> {
    type Out;
}

impl<Env, T> Eval_<Env> for T {
    default type Out = T;
}

#[test]
fn test() {
    type List = Cons<(usize, i8), Cons<(bool, i16), Cons<((), i32), Nil>>>;
    type I16 = Lookup<bool, List>; // should be i16
    let x: I16 = 1; // should compile, but gets "expected associated type". For some reason the associated type is not expanded.
}
```

and I get

```
src/lib.rs:148:18: 148:19 error: mismatched types:
 expected `<Cons<(usize, i8), Cons<(bool, i16), Cons<((), i32), Nil>>> as Lookup_<bool>>::Value`,
    found `_`
(expected associated type,
    found integral variable) [E0308]
src/lib.rs:148     let x: I16 = 1; // should compile, but gets "expected associated type". For some reason the associated type is not expanded.
                                ^
src/lib.rs:148:18: 148:19 help: run `rustc --explain E0308` to see a detailed explanation
```

**Comment by dylanede** [comment]
Here's a concise example of a problem I've found.

This code compiles fine:

```rust
use std::marker::PhantomData;

trait Id_ {
    type Out;
}

type Id<T> = <T as Id_>::Out;

impl<T> Id_ for T {
    type Out = T;
}

fn coerce<A>(_: PhantomData<A>, x: A) -> A { x }

#[test]
fn test() {
    type Bool = Id<bool>;
    let x = coerce(PhantomData::<Bool>, true);
}
```

But this, with the only difference being the "default" keyword, does not:

```rust
use std::marker::PhantomData;

trait Id_ {
    type Out;
}

type Id<T> = <T as Id_>::Out;

impl<T> Id_ for T {
    default type Out = T;
}

fn coerce<A>(_: PhantomData<A>, x: A) -> A { x }

#[test]
fn test() {
    type Bool = Id<bool>;
    let x = coerce(PhantomData::<Bool>, true);
}
```

The error reported is "associated type expected, found bool" in the second argument to the call to `coerce`.

**Comment by aturon** [comment]
@dylanede Yep, looks like a bug. I'm a bit slammed with other things but I'll try to add this as an additional test for this PR :)

**Comment by dylanede** [comment]
Another thing I've noticed. In some cases if a specialization of an `impl` appears earlier in the file than the `impl` it specializes, I get an ICE with "unwrap on a None".

**Comment by dylanede** [comment]
For example, this code gets the ICE:

```rust
trait Foo {
    type Out;
}

impl Foo for bool {
    type Out = ();
}

impl<T> Foo for T {
    default type Out = bool;
}
```

**Comment by dylanede** [comment]
Another, different ICE:

```rust
trait Id_ {
    type Out;
}

type Id<T> = <T as Id_>::Out;

impl<T> Id_ for T {
    default type Out = T;
}

fn main() {
    let x: Id<bool> = panic!();
}
```

Error and backtrace:
```
error: internal compiler error: type_of with TyProjection
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
thread 'rustc' panicked at 'Box<Any>', src/libsyntax/errors/mod.rs:230
stack backtrace:
   1:     0x7fd80e171460 - sys::backtrace::tracing::imp::write::h51a4398742da0623K0t
   2:     0x7fd80e177fa5 - panicking::log_panic::_<closure>::closure.41506
   3:     0x7fd80e177a67 - panicking::log_panic::hf3690f0da42f7f25Y7x
   4:     0x7fd80e141043 - sys_common::unwind::begin_unwind_inner::h53054637973a059djTs
   5:     0x7fd80a1d4657 - sys_common::unwind::begin_unwind::begin_unwind::h16847477020784951041
   6:     0x7fd80a1d4a14 - errors::Handler::bug::h623de609723e663bAZc
   7:     0x7fd80b5fd4ec - session::Session::bug::hf2ec81b68b650bda9es
   8:     0x7fd80cae9c7d - trans::type_of::in_memory_type_of::hb00ac4f71a5d2dc2bdQ
   9:     0x7fd80caed9cb - trans::base::alloc_ty::h4a46066a01d64342yhi
  10:     0x7fd80cc08f54 - trans::_match::mk_binding_alloca::h16584621283196570395
  11:     0x7fd80cb0986e - trans::base::init_local::h3cc56120bc50056al2h
  12:     0x7fd80cb23e12 - trans::controlflow::trans_block::h14e00c38eb84fe83VAw
  13:     0x7fd80cb204f7 - trans::base::trans_closure::hd70f9863156f5359bQi
  14:     0x7fd80cb2478c - trans::base::trans_fn::hf204ab03c9603db1XZi
  15:     0x7fd80cb28caf - trans::base::trans_item::hc86e02775fa6daedXnj
  16:     0x7fd80cb34c64 - trans::base::trans_crate::h1366c98d49e7fc59K7j
  17:     0x7fd80e6e50a1 - driver::phase_4_translate_to_llvm::h72f03170468523ffuTa
  18:     0x7fd80e6def36 - driver::phase_3_run_analysis_passes::_<closure>::closure.25890
  19:     0x7fd80e6bcc05 - middle::ty::context::ctxt<'tcx>::create_and_enter::create_and_enter::h12957577696868225661
  20:     0x7fd80e6b8521 - driver::phase_3_run_analysis_passes::h14675146087569771906
  21:     0x7fd80e68ceb9 - driver::compile_input::hcea038a856e2e417jca
  22:     0x7fd80e67f0cb - run_compiler::h6bc599602c13f1ce5wc
  23:     0x7fd80e67be86 - sys_common::unwind::try::try_fn::try_fn::h12460656089999964736
  24:     0x7fd80e16f0f8 - __rust_try
  25:     0x7fd80e1668eb - sys_common::unwind::try::inner_try::h193a98f8a6ded06bRPs
  26:     0x7fd80e67c1e0 - boxed::F.FnBox<A>::call_box::call_box::h4438575142861133636
  27:     0x7fd80e1764b3 - sys::thread::Thread::new::thread_start::he0b127204317aff8bax
  28:     0x7fd807a430a3 - start_thread
  29:     0x7fd80ddf504c - clone
  30:                0x0 - <unknown>
```

If the `default` keyword is removed it compiles fine.

**Comment by dylanede** [comment]
Similarly, if I try `std::instrinsics::type_name` on any type retrieved from an associated type of a specialized trait impl, I get an ICE complaining that the type is unexpected, where the type shown is the unnormalized associated type.

**Comment by alexcrichton** [comment]
:boat:

**Comment by aturon** [comment]
@dylanede BTW, are you following the thread on the RFC? We are [moving toward removing specialization of associated types](https://github.com/rust-lang/rfcs/pull/1210#issuecomment-169886684).

**Comment by aturon** [comment]
@dylanede (In particular, I'm now in the process of rebasing and addressing the initial review, and plan to remove the code for associated type specialization.)

**Comment by dylanede** [comment]
@aturon Ah, I hadn't been. Bear in mind that if you remove specialisation of associated types, the metaprogramming capabilities are pretty much completely removed. As it is I don't think Rust's type system is Turing Complete at the moment (ever since the changes that added the rule that type parameters must be in the trait being impl'ed or the type being impl'ed for) - I have repeatedly tried and failed to implement the untyped lambda calculus in it. With proper associated type specialisation, the implementation of ULC is almost trivial. I currently have prototype libraries waiting on better metaprogramming support, as the type system as it is doesn't seem capable enough - an Earley-based CFG parsing library and an expression-template style linear algebra library with similar aims to Eigen for C++.

**Comment by aturon** [comment]
@dylanede So, part of the debate on the RFC is whether, given the constraints we have to put on projections, specializable associated types are all that useful. If you could leave a comment on thread giving some concrete details of your use case, that'd be very helpful!

**Comment by drbo** [comment]
As I understand it, @dylanede's use case is type-level metaprogramming.  This allows for design patterns currently not possible in Rust, such as [expression templates][1].  Expression templates can achieve _massive_ performance improvements (they are analogous to stream fusion in Haskell).

[1]: https://en.wikipedia.org/wiki/Expression_templates

**Comment by eddyb** [comment]
> Expression templates can achieve _massive_ performance improvements (they are analogous to stream fusion in Haskell).

Keep in mind that stream fusion is unnecessary in Rust, as lazy lists (which can be inadvertently fully evaluated) were replaced by "strictly lazy" streams (`Iterator`), effectively "fused" at monomorphization time.
LLVM only really needs to remove abstraction boundaries to get the exact same performance as the equivalent loop.

So I'd take "massive" with a grain of salt, but that's just me.

**Comment by aturon** [comment]
@nikomatsakis 

I've now rebased and pushed a commit that addresses the more basic nits you brought up.

There are several more things that need to happen before landing:

- I want to resolve the nits around walking up the specialization graph by exploring an iterator-based approach, as you suggested; I'm hopeful this can also reduce a bit of code duplication.

- I have not had success trying to follow better hygiene with "orphaned" substs and so on (that are floating around independently of any `InferCtxt`, but might have inference variables within). The code is carefully written so that it doesn't matter, but I agree with @nikomatsakis that this creates opportunities for future bugs. 

  - The main problem I'm hitting is an inability to either (1) update the parameter environment in a reasonable way or (2) clone an inference context. That's forcing me to gin up new contexts out of thin air, which happens to work for my purposes here but is error prone.

- There are a handful of TODOs. Most of them are just me checking in with @nikomatsakis about things I'm not sure about. But we also need to e.g. fully work out the story with "auto" traits and negative impls.

- I want to track down the ICEs and weird limitations that @dylanede discovered.

I'm hoping to have all these addressed before the RFC is accepted, but @nikomatsakis I need your help on several of these points.

**Comment by aturon** [comment]
Oh, also, my branch is currently failing some dependency graph tests, and I don't fully understand why.

**Comment by aturon** [comment]
@nikomatsakis 

Update: this PR is ready for re-review. At this point the commit history is a bit gnarly, so it may make the most sense to just look directly at the diff. But if you prefer I can highlight the most important new commits.

Some of the things I've done in this latest push:

- Cleaned up all cases of inference variables potentially floating around without their context. There was a key insight here: when we need to translate between substs (going from a specialized impl to a more general one to pull defaults during projection or trans), *we already have an inference context that can successfully prove the bounds for the more specialized impl*. Previously, the code was trying to construct a new context where these bounds were added as assumptions. Now, the code just threads through the original inference context.

- Rewrote the various specialization graph traversals to provide explicit iterators and various helpers for locating items by name.

- Disallowed specialization to change impl polarity (for `auto` traits).

- Fixed the worst of the ICEs pointed out by @dylanede.

Only some minor issues remain:

- There are a small number of TODO items, some of which are questions we still need to work through.
- There are some deeper ICEs of @dylanede that I need to examine -- most likely to do with normalization.
- I need to do a bit more work on tests, and in particular figure out why the code had broken some dependency graph tests -- maybe you can help me with the latter.

**Comment by nikomatsakis** [comment]
:+1: will look soon

**Comment by aturon** [comment]
Aside: the snapshot registration commit will be supplanted by https://github.com/rust-lang/rust/pull/31778

**Comment by aturon** [comment]
Update: fixed another of @dylanede's ICEs, and along the way refactored so that the "projection mode" is a part of InferContext rather than SelectionContext, hopefully making it easier to ensure that projection knows when it's in "trans mode".

**Comment by eddyb** [comment]
@aturon Sounds like I could reuse that for deanonymization of `impl Trait` :wink:.

**Comment by aturon** [comment]
@eddyb Makes perfect sense -- it's basically the identical situation.

**Comment by m4rw3r** [comment]
Hi, I have been looking at this feature with interest since it would hopefully also open up for some more generic programming. Today I decided to try it out (since the RFC landed and all) and checked out 7e76e87 and compiled it.

After some testing I ran into a somewhat odd issue regarding traits with type-parameters. It might be that I am misunderstanding some of the rules in the RFC, but I am a bit confused as to why the following does not work (I am getting a conflict here):

```rust
struct Data;

trait Foo<T> {}
trait Bar<T> {}

impl<T, U> Bar<U> for T    where T: Foo<U> {}
impl<U>    Bar<U> for Data {}
```
while the following will work:

```rust
struct Data;

trait Foo {}
trait Bar {}

impl<T> Bar for T    where T: Foo {}
impl    Bar for Data {}
```

Wouldn't the specific `impl<U> Bar<U> for Data` be more specific than the generic `impl` for `T: Foo<U>`?

**Comment by arielb1** [comment]
One teeny little issue:
```Rust
#![feature(specialization, type_ascription)]

trait Coerce<U> {
    fn coerce(self) -> Option<U>;
}

impl<U,V> Coerce<U> for V {
    default fn coerce(self) -> Option<U> { None }
}

impl<U> Coerce<U> for U {
    fn coerce(self) -> Option<U> { Some(self) }
}

fn coerce<V, U>(v: V) -> Option<U> {
    <V as Coerce<U>>::coerce(v)
}

fn main() {
    let a = String::from("hello");

    if let Some(bomb) = coerce(&a): Option<&String> {
        drop(a);
        let _ = String::from("0wned");
        println!("{}", bomb);
    }
}
```

This crashes pretty nicely.

I did talk about this in the RFC - this obviously has to be fixed somehow.

**Comment by aturon** [comment]
@m4rw3r 

The scenario you're describing isn't actually a specialization: there's no particular relationship between `Data` and `T: Foo`. The reason that the version without generics works is that the compiler can see that there's no overlap, because `Data` doesn't impl `Foo`, and everything involved is crate-local. But with type parameters, downstream crates could impl `Foo<T>` for their own types. See the [rebalancing coherence RFC](https://github.com/rust-lang/rfcs/pull/1023) for more details.

(In general, specialization works when there *is* overlap, but one impl is "contained within" another -- roughly, the `where` clause for one impl implies everything in the `where` clause for the other. The RFC gives a lot more detail.)

@arielb1 

Indeed, as I say at the top of the PR, there are several known gaps, including:

> no checking for lifetime-dependent specialization (a soundness hole);

As per discussion with @nikomatsakis, I plan to work on this issue in a separate, follow-up PR.

**Comment by m4rw3r** [comment]
@aturon Thank you for the links, I can see the issue now. And the documents about negative reasoning even says that "Specialization is a very powerful technique that almost helps with this problem, but in fact does not.". Slightly disappointing that such a construction is not possible at the moment since it would open up for some interesting things.

**Comment by nikomatsakis** [comment]
OK, this PR is, by and large, a thing of beauty. Much improved since its early incarnations. I wound up just checking out the code, so I pushed a [commit](https://github.com/nikomatsakis/rust/commit/1ff8ee344b1d146f67516ae73277cc2c5d294bb3) to my repo containing nits and other small changes I made as I went along. 

I went ahead and moved the test files into their own subdirectory, but you need to rebase I think to get the improvements to compiletest that support that. You may also be interested in the revisions added by https://github.com/rust-lang/rust/pull/32007, though I didn't see any obvious place that you would want them.

Some general comments about tests: first, I highly recommend you give at least SOME error string. I've found something minimal like an `E0123` code or else something like `"mismatched types"` to be good. You don't want too much text, because that means you have to update the tests all the dang time. But too little and you can easily wind up with random errors (like typos) that make you think your tests are working when in fact they are not.

Second, I've been wanting to establish some new conventions that allow us to keep a better handle on what scenarios are tested and which are not. I am wondering if we can make a `README.md` file in the `specialization` test directories that tries to summarize the kinds of scenarios that arise and where they are tested? I'm not really sure how such a file would look though so feel free to ignore me -- but, basically, if you have tried to think rigorously about what kinds of situations can arise, I would encourage you to transpose those notes into a file. Otherwise, I will try to do so tomorrow maybe.

There were a few comments in that commit that are not nits. Let me know what you think about those questions! Notably:

- [ ] https://github.com/nikomatsakis/rust/commit/1ff8ee344b1d146f67516ae73277cc2c5d294bb3#diff-495b73ec30324c8a1f8c724d5fb7e258R142
- [ ] https://github.com/nikomatsakis/rust/commit/1ff8ee344b1d146f67516ae73277cc2c5d294bb3#diff-495b73ec30324c8a1f8c724d5fb7e258R209

Also, I didn't answer your TODOs about regions. I ran out of time to think about that, but maybe we can touch base on this tomorrow. :) The answer seems connected to the matter of lifetime dispatch etc.

**Comment by aturon** [comment]
@nikomatsakis Pushed an update that addresses all of your feedback, except for adding a `README` to the tests (which I'll do shortly). Take a look at the "Address review comments" commit and see what you think.

Otherwise, here's what remains:

- Add a few more tests (marked as TODO in source);
- Add README for new test directories;
- Track down @dylanede's [bug](https://github.com/rust-lang/rfcs/pull/1210#issuecomment-189619825) -- the projection *should* be allowed here (since the type is not `default`), but it should unify with `bool` in all the circumstances raised in the comments.

Only the first one is a blocker for landing.

In any case, @nikomatsakis, I'm hoping you can sign off on the review commit and give me r+, and then I'll take care of these last items and send to bors.

**Comment by aturon** [comment]
(Note, the above list of course doesn't include the bits considered out of scope for this PR, which are listed in the PR description.)

**Comment by nikomatsakis** [comment]
r=me modulo latest comments

**Comment by dylanede** [comment]
Cross-posting from the RFC issue for visibility (it's probably more suited to this issue anyway):

@aturon @nikomatsakis 
Here's something to think about. This is using the current implementation. This shows a compile-time dependency on the hidden type of a `default`ed associated type. Whether this is relying on behaviour not valid according to the RFC is another matter.

```rust
fn main() {
    enum Undefined {}
    trait Opaque {
        type Out;
    }
    impl<T> Opaque for T {
        default type Out = Undefined;
    }
    impl Opaque for () {
        type Out = bool;
    }

    trait Same<A, B> {
        type T;
        fn cast(a: A) -> B;
    }
    impl<A> Same<A, A> for () {
        type T = A;
        fn cast(a: A) -> A {
            a
        }
    }
    fn cast<X, Y>(x: X) -> Y where (): Same<X, Y> {
        <() as Same<X, Y>>::cast(x)
    }

    // Bool is an "opaque" type that is secretly bool
    // This will not work: let x: Bool = true;
    type Bool = <() as Opaque>::Out;

    // But the following does work:

    // compilation dependency:
    {
        fn foo<X, Y>(y: Y) where (): Same<X, Y> {

        }
        let x: bool = true;
        foo::<Bool, _>(x); // only compiles when Bool is bool
        // also, if Bool is not bool, the compilation error message will reported the normalised type
    }
    // getting a bool from a Bool works
    {
        let x: Bool = unimplemented!(); // a Bool
        let x: bool = cast(x); // x is now a bool, at compile time, requires Bool == bool
    }
    // getting a Bool from a bool won't compile though
    {
        let x: bool = true;
        let x: Bool = cast(x); // doesn't compile
    }
}
```

**Comment by nikomatsakis** [comment]
@dylanede is this the same example you posted on the rust-lang/rfcs#1210? If so, there is a bug, but it is not the compilation time dependency -- it's rather the error on the last line, which is a bit perplexing (likely a missing normalization).

It is expected that you can project `<() as Opaque>::Out>` to `bool` -- this is because the impl for `()` defined `type Out = bool` without the keyword `default`. What would be a bug is if the keyword `default` were used there.

The case that we were debating was this one:

- If the most specific impl for the type you have uses `default` (not the case in your example);
- but there cannot be a more specific impl (because of orphan rules);
- should we allow you to rely on the type or not?

The answer to *that* question should be no.

So, in your example, if I project `<u16 as Opaque>::Out` I should *not* get `Undefined` (or at least should not be able to rely on it).

**Comment by aturon** [comment]
@dylanede 

Turns out the problem you raised is actually an existing normalization bug for type aliases, unrelated to specialization:

```rust
fn main() {
    enum Undefined {}
    trait Opaque {
        type Out;
    }
    impl Opaque for () {
        type Out = bool;
    }
    type Bool = <() as Opaque>::Out;
    let x: <() as Opaque>::Out = true; // works
    let x: Bool = true; // fails
}
```

**Comment by aturon** [comment]
@nikomatsakis Hm, no clue about the [test failure](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/8403/steps/test/logs/stdio):

```
---- ext::expand::tests::item_macro_workaround stdout ----
	thread 'ext::expand::tests::item_macro_workaround' panicked at 'idents with different contexts are compared with operator `==`: fn#3, default#0.', ../src/libsyntax/ast.rs:121
```

Will try to look into it tomorrow...

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -644,6 +644,22 @@ impl<'a> Parser<'a> {
         }
     }
 
+    pub fn check_contextual_keyword(&mut self, ident: Ident) -> bool {
+        let tok = token::Ident(ident, token::Plain);
+        let r = self.token == tok;
```
Comment (Aatch) [inline src/libsyntax/parse/parser.rs]: This is probably wrong, see the comment for the impl of `Eq` for `Ident` (ast.rs:104). You probably want non-hygenic comparison here.

**Comment by alexcrichton** [comment]
:tanabata_tree:

**Pull request merged**